src/authenticator.rs

```diff
@@ -698,6 +689,114 @@ where
+    ///
+    /// https://github.com/Nitrokey/nitrokey-hotp-verification#verifying-hotp-code
+    /// Solution contains a mean to avoid desynchronization between the host's and device's counters. Device calculates up to 9 values ahead of its current counter to find the matching code (in total it calculates HOTP code for 10 subsequent counter positions). In case:
+    ///
+    /// - no code would match - the on-device counter will not be changed;
+    /// - incoming code parsing would fail - the on-device counter will not be changed;
+    /// - code would match, but with some counter's offset (up to 9) - the on-device counter will be set to matched code-generated HOTP counter and incremented by 1;
+    /// - code would match, and the code matches counter without offset - the counter will be incremented by 1.
+    ///
+    /// Device will stop verifying the HOTP codes in case, when the difference between the host and on-device counters will be greater or equal to 10.
+    ///
+    /// ```
+    ///
+    /// ```
+    fn verify_code<const R: usize>(&mut self, args: VerifyCode, reply: &mut Data<{ R }>) -> Result {
+        const COUNTER_WINDOW_SIZE: u32 = 9;
+        // TODO DESIGN allow to use any credential this way? or limit usage to this operation during credential creation?
+        let credential = self.load_credential(&args.label).ok_or(Status::NotFound)?;
+
+        // TODO move to u64 integer
+        // Convert the incoming code from string to u32.
+        // No need for the zero left-pad, or alloc for the format! macro this way.
+        if args.response.len() > credential.digits as usize {
+            return Err(Status::ConditionsOfUseNotSatisfied);
+        }
+        // Validate input data without using up attempt
+        let code_in = convert_string_to_integer(args.response)?;
+
+
+        let current_counter = match credential.kind {
+            oath::Kind::Totp => return Err(Status::ConditionsOfUseNotSatisfied),
+            oath::Kind::Hotp => {
+                if let Some(counter) = credential.counter {
+                    counter
+                } else {
+                    debug_now!("HOTP missing its counter");
+                    return Err(Status::UnspecifiedPersistentExecutionError);
+                }
+            }
+        };
+        let mut found = None;
+        for offset in 0..=COUNTER_WINDOW_SIZE {
+            // TODO Do abort with error on the max value, so these could not be pregenerated,
+            // and returned to user after overflow, or the same code used each time
+            let counter = current_counter.saturating_add(offset);
+            let code = self.calculate_hotp_code_for_counter(credential, counter);
+            if code == code_in {
+                found = Some(counter);
+                break;
+            }
+        }
+
+        if found.is_none() {
+            // Failed verification
+            // TODO blink red LED infinite times, highest priority
```

Comment:
This TODO seems important.

This is blocked by trussed only offering and "Wink" without more configuration right?

Reply:
Correct. Leaving at blinking at the moment. To be updated later.
- https://github.com/Nitrokey/trussed/issues/5